GVFS/GVFS.Service/RepoAutoMounter.cs

```diff
@@ -0,0 +1,150 @@
+            {
+                foreach (RepoRegistration repo in activeRepos)
+                {
+                    if (rootPath == null || GVFSPlatform.Instance.FileSystem.IsPathUnderDirectory(rootPath, repo.EnlistmentRoot))
+                    {
+                        this.Mount(repo.EnlistmentRoot);
+                    }
+                }
+            }
+            else
+            {
+                this.tracer.RelatedError("Could not get repos to auto mount for user. Error: " + errorMessage);
+            }
+        }
+
+        private void Mount(string enlistmentRoot)
+        {
+            var metadata = new EventMetadata
+            {
+                ["EnlistmentRoot"] = enlistmentRoot
+            };
+
+            using (var activity = this.tracer.StartActivity("AutoMount", EventLevel.Informational, metadata))
+            {
+                string volumeRoot = GVFSPlatform.Instance.FileSystem.GetVolumeRoot(enlistmentRoot);
+                if (GVFSPlatform.Instance.FileSystem.IsVolumeAvailable(volumeRoot))
+                {
+                    // TODO #1043088: We need to respect the elevation level of the original mount
+                    if (this.mountProcessManager.StartMount(enlistmentRoot))
+                    {
+                        this.SendNotification("GVFS AutoMount", "The following GVFS repo is now mounted:\n{0}", enlistmentRoot);
+                        activity.RelatedInfo("Auto mount was successful for '{0}'", enlistmentRoot);
+                    }
+                    else
+                    {
+                        this.SendNotification("GVFS AutoMount", "The following GVFS repo failed to mount:\n{0}", enlistmentRoot);
+                        activity.RelatedError("Failed to auto mount '{0}'", enlistmentRoot);
+                    }
+                }
+                else
+                {
+                    activity.RelatedInfo("Cannot auto mount '{0}' because the volume '{1}' not available.", enlistmentRoot, volumeRoot);
+                }
+            }
+        }
+
+        private void OnVolumeStateChanged(object sender, VolumeStateChangedEventArgs e)
+        {
+            var metadata = new EventMetadata
+            {
+                ["State"] = e.ChangeType.ToString(),
+                ["Volume"] = e.VolumePath,
+            };
+
+            this.tracer.RelatedEvent(EventLevel.Informational, "VolumeStateChange", metadata);
+
+            switch (e.ChangeType)
+            {
+                case VolumeStateChangeType.VolumeAvailable:
+                    this.MountAll(rootPath: e.VolumePath);
+                    break;
+                case VolumeStateChangeType.VolumeUnavailable:
+                    // There is no need to do anything here to stop any potentially orphaned mount processes
+                    // since they will self-terminate if the volume is removed.

```

Comment:
This comment lies. Mount processes do not self-terminate.. but should they? Open question 1.

Reply:
They do self-terminate the next time they try to do some IO and it fails.